src/main/java/com/aws/greengrass/lifecyclemanager/KernelAlternatives.java

```diff
@@ -392,6 +417,91 @@ public void cleanupLaunchDirectoryLinks() {
+        try {
+            rollbackConfig.read(deploymentDirectoryManager.getSnapshotFilePath());
+        } catch (IOException exc) {
+            logger.atError().log("Failed to read rollback snapshot config", exc);
+            return false;
+        }
+        boolean bootstrapOnRollbackRequired = false;
+        try {
+            // Check if we need to execute component bootstrap steps during the rollback deployment.
+            // Exclude components with bootstrap steps that did not execute during the target deployment.
+            final Set<String> componentsToExclude = bootstrapManager.getPendingTasks();
+            logger.atDebug().kv("components", componentsToExclude)
+                    .log("These components did not bootstrap during the target deployment. "
+                            + "They will be excluded from bootstrap-on-rollback.");
+            // Exclude components that are not explicitly configured to bootstrap-on-rollback
+            Set<String> unconfiguredComponents = getComponentsNotConfiguredToBootstrapOnRollback(rollbackConfig);
+            logger.atDebug().kv("components", unconfiguredComponents)
+                    .log("These components are not configured to execute bootstrap steps during rollback. "
+                            + "They will be excluded from bootstrap-on-rollback.");
+            componentsToExclude.addAll(unconfiguredComponents);
+            bootstrapOnRollbackRequired = bootstrapManager.isBootstrapRequired(rollbackConfig.toPOJO(),
+                    componentsToExclude);
+        } catch (ServiceUpdateException | ComponentConfigurationValidationException exc) {
+            logger.atError().log("Rollback config invalid or could not be parsed", exc);
+            return false;
+        }
+        Path rollbackBootstrapTaskFilePath;
+        try {
+            rollbackBootstrapTaskFilePath = deploymentDirectoryManager.getRollbackBootstrapTaskFilePath();
+        } catch (IOException exc) {
+            logger.atError().log("Bootstrap-on-rollback task file path could not be resolved", exc);
+            return false;
+        }
+        if (bootstrapOnRollbackRequired) {
+            // Bootstrap-on-rollback is required, so write the task file.
+            try {
+                bootstrapManager.persistBootstrapTaskList(rollbackBootstrapTaskFilePath);
+            } catch (IOException exc) {
+                logger.atError().log("Bootstrap-on-rollback task file could not be written", exc);
+                return false;
+            }
+        } else {
+            // Bootstrap-on-rollback is not required, so ensure that the task file is deleted.
+            try {
+                bootstrapManager.deleteBootstrapTaskList(rollbackBootstrapTaskFilePath);
+            } catch (IOException exc) {
+                logger.atError().log("Bootstrap-on-rollback task file could not be cleaned up", exc);
+                return false;
+            }
```

Comment:
would this scenario happen in reality?

Reply:
I am trying to preclude possible corner cases.

During rollback, Nucleus uses the existence of the `rollback_bootstrap_tasks.json` file to determine whether to launch in deployment stage `ROLLBACK_BOOTSTRAP` vs. `KERNEL_ROLLBACK`.

I'm imagining a scenario where a rogue `rollback_bootstrap_tasks.json` gets saved into the file system, as a result of some improper Nucleus upgrade/downgrade. In that case, the best thing to do is to pro-actively delete the file (if it exists) when we know that `ROLLBACK_BOOTSTRAP` is not the desired next deployment stage.